teps/0037-simplify-metrics.md

```diff
@@ -0,0 +1,230 @@
+While single users who care about statistics of individual `TaskRun` can still keep the fine-grained metrics.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Coarse-grained metrics should be provided to satisfy the users' needs. This can be implemented in 2 ways: changing `TaskRun` level metrics to namespace level and changing metrics type. These 2 methods are mutually exclusive.
+If users care about individual `TaskRun`'s performance, they can set `duration_seconds` type from histogram to gauge. If users care about overall performance, they can collect metrics at namespace level.
+
+### Change TaskRun level metrics to namespace level
+
+We can add a `config-observability` option to switch between `TaskRun` level metrics and namespace level.
+`metrics.namespace-level` field indicates whether to aggregate metrics at namespace level. When set to `true`, it will remove `Task` and `TaskRun` label in metrics. Sizable reduction in metrics count can
+happen when multiple `TaskRun`s is running under the same namespace.
+By default, it is set to `false`.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-observability
+  namespace: tekton-pipelines
+  labels:
+    app.kubernetes.io/instance: default
+    app.kubernetes.io/part-of: tekton-pipelines
+data:
+  metrics.namespace-level: "true"
+```
+
+When the option is `false`, every `TaskRun` in a namespace has a set
+of seperate metrics to record its running status.
+
+```log
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-ytqrdxja",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-ymelobwl",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-xnuasulj",le="10"} 1
+```
+
+When the option is `true`, these metrics will be merged into one.
+
+```log
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",le="10"} 3
+```
+
+### Change metrics type
+
+Large metrics volume is mostly caused by histogram metrics. One single histogram metrics will produce 15 metrics. These metrics type could be changed to gauge which would reduce 15 metrics to one.
+For example, metrics like `tekton_pipelinerun_duration_seconds`, `taskrun_duration_seconds`, `tekton_pipelinerun_taskrun_duration_seconds` are histogram. But the metrics could not provide much information at `TaskRun` level and can be changed to gauge.
```

Comment:
is it possible to explain or link to what "histogram" means? (maybe we can add a link to https://github.com/tektoncd/pipeline/blob/master/docs/metrics.md)

Reply:
This looks like a good definition: https://prometheus.io/docs/concepts/metric_types/

Reply:
They would make sense as an histogram once they are aggregated over a namespace or across namespaces, right? Would we have the metric as a gauge on taskrun level and as histogram at higher levels of aggregation?